EIPS/eip-5843.md

```diff
@@ -0,0 +1,448 @@
+#
+# input:
+#	x, y, mod - multiprecision numbers represented by arrays of 64bit unsigned integers in ascending order of significance.
+#	x, y < mod
+#	mod % 2 != 0
+#	mod_inv_small = pow(-mod, -1, 1<<64)
+# output:
+#	(x * y * pow(R, -1, mod)) % mod represented by array of 64bit unsigned integers in ascending order of significance
+#
+def mulmont_cios(x: [int], y: [int], mod: [int], mod_inv_small: int) -> [int]:
+    assert len(x) == len(y) and len(y) == len(mod), "{}, {}, {}".format(x, y, mod)
+    assert mod[-1] != 0, "modulus must occupy all limbs"
+    limb_count = len(mod)
+    t = [0] * (limb_count + 2)
+
+    # each outer-loop iteration computes a row of the multiplication (x * y) % R
+    # and simultaneously adds a factor to make the intermediate value divisible by (1<<64)**i
+    # until after the final iteration, T is divisible by R
+
+    for i in range(limb_count):
+        # first inner-loop multiply x * y[i]
+        c = 0
+        for j in range(limb_count):
+            c, t[j] = hi_lo(t[j] + x[j] * y[i] + c)
+
+        # TODO why 2 carry words instead of 1? Has to do with saving an extra word because we interleave the shift?
+        t[limb_count + 1], t[limb_count] = hi_lo(t[limb_count] + c)
+
+        # second inner-loop: t <- t + ((t % b) * modinv % b) * mod making t % b**i == 0
+        # integrates the reduction (bit-shift) into the calculation of t
+        m = (mod_inv_small * t[0]) % BASE
+        c, _ = hi_lo(m * mod[0] + t[0])
+
+        for j in range(1, limb_count):
+            c, t[j - 1] = hi_lo(t[j] + mod[j] * m + c)
+
+        # TODO explain carries here
+        hi, t[limb_count - 1] = hi_lo(t[limb_count] + c)
+        t[limb_count] = t[limb_count + 1] + hi
+
+    # discard the top-limb.  the result is 0 < result < 2 * mod so it can't overflow two limbs
+    t = t[:-1]
+    if t[-1] != 0:
+        # result is greater than R, and also greater than mod
+        return limbs_sub(t, mod + [0])[:-1]
+    elif limbs_gte(t[:-1], mod):
+        return limbs_sub(t[:-1], mod)
+    else:
+        return t[:-1]
+```
+
+TODO: demonstrate that `mulmont_cios` has the same properties as `mulmont`.
+
+### 2 - Modular Inverse - Extended Euclidean Algorithm
+
+The Extended Euclidean algorithm takes two numbers `x`,`y` as input and outputs `a`,`b`,`v` such that `a*x + b*y = v = gcd(x,y)`, `a` and `b` are numerically the smallest values that satisfy the relationship.  When `x` and `y` are coprime, the algorithm can be used to find `pow(x, -1, y)`:
+
+```
+a*x + by = 1
+a*x - 1  = (-b)*y 
+a*x - 1  = 0 % y
+a*x      = 1 % y
+a       = pow(x, -1, y)
+```
+
+```
+def extended_euclidean_algorithm():
+	pass #TODO
+```
+
+TODO:
+* Some derivation of complexity for extended euclidean (https://www.scaler.com/topics/data-structures/extended-euclidean-algorithm/ claims `O(log(min(x, y))`.
+* investigate go standard library which uses Lehmer's algorithm as part of calculating inverse.
```

Comment:
The bigint modular inverse is not needed for addmodx or mulmodx.
The only inverse that is needed fits in a word and it's cheap to compute it.

Now if an EIP needs full blown modular inverse, Bernstein-Yang and Pornin's GCDs from 2019 and 2020 are significantly faster (like 5x to 10x) than any other algorithm like Lehmer's, see https://github.com/mratsim/constantine/pull/172

Reply:
The "full" `mod_inv` is needed for generic `mulmont` which is used at larger bit-widths.  I think it would be sensible to modify the spec to not compute it if the bitwidth is below a certain cutoff.

Reply:
Here are 2 implementations of mulmont using CIOS and FIPS, mod_inv is not needed:

https://github.com/mratsim/constantine/blob/928f5155827252c01ee0c7a5c58c8b9deecc928f/constantine/math/arithmetic/limbs_montgomery.nim#L217-L307

m0ninv is the `-1/M[0]` the negative modular multiplicative inverse of the first limb of the prime modulus.

The way modular reduction works is by multiplying with it you can zero put the bottom bits and then shift by a whole word, one word at a time, you don't need to multiply by the full inverse.